### PR TITLE
Adjust signatures of eigenvalues and eigenvectors to accept expressions

### PR DESCRIPTION
**Pull request by WardBrian**
## Summary

This adjusts the signatures of the `eigenvalues` and `eigenvectors` functions to accept eigen expressions as input. This will allow them to be exposed in the Stan language: https://github.com/stan-dev/stanc3/pull/1192

## Tests

Existing tests for functionality remain in place. Additional expression tests will be automatically run when exposed in the compiler, I've run them locally for now.

## Side Effects

None

## Release notes

Updated the signatures of `eigenvalues` and `eigenvectors` to use our pseudo-concept require templates. 

## Checklist

- [ ] Math issue #(issue number)

- [x] Copyright holder:  Simons Foundation

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/prim/fun/eigenvalues.hpp [hunk 1]
```diff
@@ -7,13 +7,14 @@
...
 namespace math {
 
-template <typename T>
-Eigen::Matrix<std::complex<T>, -1, 1> eigenvalues(
-    const Eigen::Matrix<T, -1, -1>& m) {
-  check_nonzero_size("eigenvalues", "m", m);
-  check_square("eigenvalues", "m", m);
+template <typename EigMat, require_eigen_matrix_dynamic_t<EigMat>* = nullptr>
+inline auto eigenvalues(const EigMat& m) {
+  using PlainMat = plain_type_t<EigMat>;
```
Comment (bob-carpenter) [inline stan/math/prim/fun/eigenvalues.hpp:12]: [question]
Is there a reason to specify `PlainMat` rather than just using `.eval()`?

```cpp
auto& m_eval = m.eval();
```

File: test/statement_types.py [hunk 2]
```diff
@@ -130,7 +130,10 @@ def __init__(self, overload, name, stan_arg, size, value=None):
         self.size = size
 
         if value is None:
-            self.value = 0.4
+            if 'complex' in stan_arg:
```
Comment (bob-carpenter) [inline test/statement_types.py:133]: [optional]
I prefer ternary operators when the body of conditionals only assign to the same variable.  But ternary ops look pretty awkward in Python (of course this one wouldn't even be legal in C++ with its heterogeneous return types), so I'm marking as optional.

```python
self.value = to_complex(0.4, 04) if 'complex' in stan_arg else 0.4
```

**Review by bob-carpenter — APPROVED** [review]
This is fine as is.  My comments were more questions/optional and about style, not content.

File: stan/math/prim/fun/eigenvalues.hpp (hunk 1 above)
Comment (WardBrian) [reply stan/math/prim/fun/eigenvalues.hpp:12]: We use `PlainMat` later as a template argument to the eigen solver, but to be completely honest I was just using the style also used in eigenvalues_sym.hpp and elsewhere. @SteveBronder - is there a reason to prefer one to the other?

File: test/statement_types.py (hunk 2 above)
Comment (WardBrian) [reply test/statement_types.py:133]: I agree in general but quite dislike the Python ternary operator's ordering (I'm much more used to the order `conditional - true value - false value` from C, Java, etc., so I still find `true value - conditional - false value` hard to parse even after using it for years)

Comment (bob-carpenter) [reply test/statement_types.py:133]: Glad to hear it's not just me finding it hard to read.  That is why I marked it optional.

File: stan/math/prim/fun/eigenvalues.hpp (hunk 1 above)
Comment (bob-carpenter) [reply stan/math/prim/fun/eigenvalues.hpp:12]: Good point about needing to use it to parameterize the solver.  So reducing this to `auto` plus `eval` won't be enough.  So I'd just leave it like this.

File: test/statement_types.py (hunk 2 above)
Comment (rok-cesnovar) [reply test/statement_types.py:133]: +1 on the Python ternary being hard to read :)

Comment (WardBrian) [reply test/statement_types.py:133]: I think it was all to avoid making something like `then` a keyword. OCaml's if expression keeps the same order as C's ?: by having the style be `if cond then true_value else false_value`, which is probably my favorite option
